actionview/lib/action_view/cache_expiry.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module ActionView
+  class CacheExpiry
+    class Executor
+      def initialize(watcher:)
+        @cache_expiry = CacheExpiry.new(watcher: watcher)
+      end
+
+      def before(target)
+        @cache_expiry.clear_cache_if_necessary
+      end
+    end
+
+    def initialize(watcher:)
+      @watched_dirs = nil
+      @watcher_class = watcher
+      @checker = nil
+    end
+
+    def clear_cache_if_necessary
+      fs_paths = all_view_paths.select do |view_path|
+        view_path.is_a?(FileSystemResolver)
+      end
+
+      watched_dirs = fs_paths.map(&:path).sort.uniq
+      if watched_dirs != @watched_dirs
+        @watched_dirs = watched_dirs
+        @checker = @watcher_class.new([], watched_dirs) do
+          clear_cache
+        end
+        @checker.execute
```

Comment:
Would this be clearer if we extracted the watched_dirs specifics to another method? Ala:

```ruby
    def clear_cache_if_necessary
      if capture_change_to_watched_dirs
        @checker = @watcher_class.new([], @watched_dirs, &:clear_cache)
        @checker.execute
      else
        @checker.execute_if_updated
      end
    end
```

Reply:
I extracted a new method: `dirs_to_watch`. I do like keeping all the ivar assignments in this method, which I think makes the flow more obvious.